### PR TITLE
WT-10575 Fix checkpoint cursor to return only stable data

### PR DESCRIPTION
**Pull request by kommiharibabu**
Returning only visible data can lead to inconsistent results when a prepared transaction gets committed in parallel to the checkpoint with a commit timestamp less than or equal to the checkpoint's stable timestamp.

To avoid producing torn transaction results, along with the commit timestamp visibility, the data must be durable according to the checkpoint stable timestamp.

### CODE REVIEW

File: src/btree/bt_delete.c [hunk 1]
```diff
@@ -169,6 +169,7 @@ __wt_delete_page(WT_SESSION_IMPL *session, WT_REF *ref, bool *skipp)
     /* History store data are always visible. No need to check visibility. */
     if (!WT_IS_HS(session->dhandle) &&
       !__wt_txn_visible(session, addr.ta.newest_txn,
+        WT_MAX(addr.ta.newest_start_durable_ts, addr.ta.newest_stop_durable_ts),
```
Comment (quchenhao) [inline src/btree/bt_delete.c:172]: Why here we all use durable ts?

Comment (kommiharibabu) [reply src/btree/bt_delete.c:172]: In that flow, we are going to mark the page as fast truncated only if everything is visible and stable. Due to that reason, we use the durable timestamp as the check.

Comment (quchenhao) [reply src/btree/bt_delete.c:172]: Can we add a comment for that?

Comment (kommiharibabu) [reply src/btree/bt_delete.c:176]: Added a comment to explain why we use the durable timestamp for the visibility check.

**Review by quchenhao — APPROVED** [review]
lgtm

**Review by agorrod — APPROVED** [review]
Thanks Hari - this looks correct to me. It's an annoyingly large change for a corner case, but I think it's reasonable that the __wt_txn_visible API takes a time point.

I thought we had a data structure for that, but couldn't find it. I'm going to create a follow on ticket to get a WT_TIME_POINT added, so we don't need to pass this triplet of times around as separate parameters any more.

File: src/include/txn_inline.h [hunk 2]
```diff
@@ -797,7 +797,8 @@ __txn_visible_id(WT_SESSION_IMPL *session, uint64_t id)
  *     Can the current transaction see the given ID / timestamp?
  */
 static inline bool
-__wt_txn_visible(WT_SESSION_IMPL *session, uint64_t id, wt_timestamp_t timestamp)
+__wt_txn_visible(
```
Comment (bvpvamsikrishna) [inline src/include/txn_inline.h:800]: Don't we need to change the wt_txn_visible_all as well ??

Comment (kommiharibabu) [reply src/include/txn_inline.h:800]: Other than the metadata table, we always use the durable timestamp passed to the `wt_txn_visible_all` function.

Comment (kommiharibabu) [reply src/include/txn_inline.h:800]: Currently, we do not have the WT_TIME_POINT as mentioned by Alex, to add a check against future changes, I believe it is better to handle it when the WT_TIME_POINT structure is added. Opinions?

Comment (quchenhao) [reply src/include/txn_inline.h:800]: I'm in favour of that.

Comment (bvpvamsikrishna) [reply src/include/txn_inline.h:800]: sgtm

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm

File: src/include/txn_inline.h (hunk 2 above)
Comment (etienneptl) [reply src/include/txn_inline.h:800]: Probably worth adding a link to this comment in the ticket that will add the new structure or add details to the scope in the ticket's description (even better)

**Comment by kommiharibabu** [comment]
evergreen merge

File: src/include/txn_inline.h (hunk 2 above)
Comment (kommiharibabu) [reply src/include/txn_inline.h:800]: Thanks, Etienne. I created the WT-10612 ticket and linked this PR comment also.

**Pull request merged**